Support list patterns

juusom opened the pull request:
WIP

Commit:
Support list patterns

juusom commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +223,73 @@ def list_runner(output_queue)
+    return if free_slots == 0
+    keys = @redis.keys(@key)
+    keys.shuffle
+    keys = keys[0..free_slots - 1]
```
TODO: don't start threads for already existing keys

lassizci commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +237,82 @@ def list_runner(output_queue)
+      begin
+        init_list_pattern_runner if @redis.nil?
+        ensure_workers(output_queue)
+        sleep(0.1)
```
Configurable refresh interval might be in order. 0.1s feels quite frequent for listing keys.

juusom replied:
Yeah maybe. Should probably try to gather some metrics on how this actually performs because whether it's configured or hardcoded, it's just a guess otherwise. Likewise, the sensible number of items processed by a thread should probably be tested somehow.

lassizci commented on lib/logstash/inputs/redis.rb:
```diff
@@ -86,6 +103,9 @@ def register
     if @data_type == 'list' || @data_type == 'dummy'
       @run_method = method(:list_runner)
       @stop_method = method(:list_stop)
+    elsif @data_type == 'pattern_list'
```
instead of adding another `elsif`, could change this to
```
case @data_type
when 'list' || 'dummy'
  ...
when 'pattern_list'
  ..
end
```
to make it a bit more elegant.

juusom replied:
I'd keep refactorings out of this PR as anything that touches the upstream code will make this rot quicker.

lassizci replied:
Might be true, although it's couple years since the upstream has touched the code the last time.

anttipitkanen commented on lib/logstash/inputs/redis.rb:
```diff
@@ -77,6 +94,17 @@ def new_redis_instance
   end
 
+  def init_threadpool
+    @threadpool_queue_max ||= 2 * @pattern_list_threads
```
This probably doesn't need to be stored instance variable since I can't find any usage outside this method? If so, we can probably skip the variable assignment altogether and use it directly

juusom replied:
Yeah, right, this was used elsewhere at some point

anttipitkanen commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
+
+    @redis.quit rescue nil
+    @redis = nil
+    @threadpool.shutdown
```
These two could be in a separate `reset_threadpool` method if you want to be fancy. It could do the similar check that `@threadpool.nil?` like with `@redis` above and reset memoization with `@threadpool = nil` to allow `init_threadpool` to start it again

anttipitkanen commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
+    @logger.debug("Launched worker for #{key}")
+    redis = new_redis_instance
+    begin
+      (0..@pattern_list_max_items-1).each do
```
There's fancy three dot version of the range operator that gives you the same range but excludes the last value (so should be identical to the manual `-1` here)

anttipitkanen commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
+        queue_event(value, output_queue)
+      end
+    ensure
+      redis.quit rescue nil
```
`... rescue nil` is a bit evil as it catches any error, including making a typo in the variable on the left side (`rediss.quit`) which would hide broken things not so conveniently. Probably not a huge deal in the two use cases I saw here, but it's something that linters prevent nowadays.

anttipitkanen commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
+  def ensure_workers(output_queue)
+    return unless threadpool_capacity?
+    keys = @redis.keys(@key)
+    keys.shuffle
```
This doesn't seem to be doing anything as the non-bang (exclamation mark) version of this Ruby method doesn't mutate the object and instead returns a new object ( https://ruby-doc.org/core-2.4.1/Array.html#method-i-shuffle ). Instead of using the bang version, probably cleanest to chain this together as `@redis.keys(@key).shuffle.each do ...` so that there's least amount of mutations / assignments

anttipitkanen commented on lib/logstash/inputs/redis.rb:
```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
+  # private
+  def init_list_pattern_runner
+    init_threadpool
+    @redis ||= connect
```
Should we use `redis_runner` here as it seems to handle failures and reconnection? As a general thing seeing the `@redis` usage in the PR changes, I'd try to avoid touching the `@redis` instance variable directly as much as possible since it seems to be used by the other internal logic and anything we do here for our own purposes can easily conflict (or we end up doing the same things again).

juusom replied:
Good catch. The starting point for this new logic was the `list_runner` logic which is pretty different from the pubsub stuff. But could definitely use redis_runner for the main loop.

anttipitkanen reviewed (approved):
Looks good to me

Commit:
Update docs

kaisecheng committed:
remove Logstash::Pipeline

kaisecheng committed:
bump version

kaisecheng committed:
Merge pull request #84 from kaisecheng/remove-ruby-execution-engine

remove ruby pipeline dependency

kaisecheng committed:
bump version

kaisecheng committed:
fix broken test

kaisecheng committed:
Merge pull request #85 from kaisecheng/remove-ruby-execution-engine

fix bump version

Commit:
Merge remote-tracking branch 'origin/master' into list-pattern-input

jsvd committed:
[skip ci] update travis ci badge from .org to .com

kares committed:
Fix: resolve crash when commands_map is set (#86)

* Fix: do not rely on redis.client anymore
* Test: redo specs with 'real' Redis mocking
* Refactor: cleanup plugin (internals)

Co-authored-by: Ry Biesemeyer <yaauie@users.noreply.github.com>

kares committed:
Fix: better (Redis) exception handling (#89)

Redis doesn't wrap all of the errors e.g. (retriable) IOError might be raised from an IO write (#88).
With the refactored error handling we make sure these are logged and retried instead of causing pipeline crashes.

Also not all specs from the suite were run on the CI, as some are tagged with redis.
These require a real Redis server, no reason not to run them against CI as well.

Commit:
Merge remote-tracking branch 'origin/master' into list-pattern-input

Commit:
Update pattern list tests according to recent refactorings

Commit:
Fix integration tests

qikiqi commented:
This branch seems to hold the most up to date stuff from master (now main) even today.

juusom commented:
Well hello! Note btw that there's still an upstream PR open: https://github.com/logstash-plugins/logstash-input-redis/pull/81 . You might want to subscribe to it in case anyone ever picks it up 😅

stachdude commented:
Hi,

any plans to add https://github.com/logstash-plugins/logstash-input-redis/pull/81 to master in the near future ? Its been a while waiting for merge now ?
